gapic/src/main/com/google/gapid/GraphicsTraceView.java

```diff
@@ -210,59 +190,147 @@ public MainTab(Type type, Function<Composite, Control> contentFactory) {
-      List<TabInfo> left = getTabs(models.settings.leftTabs, allTabs, models, widgets);
-      List<TabInfo> center = getTabs(models.settings.centerTabs, allTabs, models, widgets);
-      List<TabInfo> right = getTabs(models.settings.rightTabs, allTabs, models, widgets);
-
-      for (Type missing : allTabs) {
-        switch (missing.defaultLocation) {
-          case Left:
-            left.add(new MainTab(missing,
-                parent -> missing.factory.create(parent, models, widgets).getControl()));
-            break;
-          case Center:
-            center.add(new MainTab(missing,
-                parent -> missing.factory.create(parent, models, widgets).getControl()));
-            break;
-          case Right:
-            right.add(new MainTab(missing,
-                parent -> missing.factory.create(parent, models, widgets).getControl()));
-            break;
-          default:
-            throw new AssertionError();
+      Iterator<String> structs = Splitter.on(';')
+          .trimResults()
+          .omitEmptyStrings()
+          .split(models.settings.tabStructure)
+          .iterator();
+      Iterator<Integer> weights = Arrays.stream(models.settings.tabWeights).iterator();
+      Iterator<String> tabs = Arrays.asList(models.settings.tabs).iterator();
+
+      FolderInfo root = parse(models, widgets, structs, weights, tabs, allTabs);
+      if (structs.hasNext()) {
+        root = null;
+      }
+      if (root == null || root.children == null) {
+        return getDefaultFolderInfo(models, widgets, hidden);
+      }
+
+      if (!allTabs.isEmpty()) {
+        TabInfo[] tabsToAdd = new TabInfo[allTabs.size()];
+        int i = 0;
+        for (Type tab : allTabs) {
+          tabsToAdd[i++] = new MainTab(tab,
+              parent -> tab.factory.create(parent, models, widgets).getControl());
         }
+        root = root.addToLargest(tabsToAdd);
       }
+      return root.children;
+    }
 
-      double[] weights = models.settings.tabWeights;
-      return new FolderInfo[] {
-          new FolderInfo(false, left.toArray(new TabInfo[left.size()]), weights[0]),
-          new FolderInfo(false, center.toArray(new TabInfo[center.size()]), weights[1]),
-          new FolderInfo(false, right.toArray(new TabInfo[right.size()]), weights[2]),
-      };
+    private static FolderInfo parse(Models models, Widgets widgets, Iterator<String> structs,
+        Iterator<Integer> weights, Iterator<String> tabs, Set<Type> left) {
+      if (!structs.hasNext() || !weights.hasNext()) {
+        return null;
+      }
+      String struct = structs.next(); // struct is non-empty (see splitter above)
+      int weight = weights.next();
+
+      int count = 0;
+      try {
+        count = Integer.parseInt(struct.substring(1));
+      } catch (NumberFormatException e) {
+        return null;
+      }
+      if (count <= 0) {
+        return null;
+      }
+
+      switch (struct.charAt(0)) {
+        case 'g':
```

Comment:
This is interpretation of a "magic string" further below, I think this should be commented a bit.